internal/config/instance.go

```diff
@@ -113,65 +118,86 @@ func Get() (*Instance, error) {
 }
 
 // Set sets a value at the given key
-func (i *Instance) Set(key string, value interface{}) {
+func (i *Instance) Set(key string, value interface{}) error {
 	i.rwLock.Lock()
 	defer i.rwLock.Unlock()
-	i.viper.Set(key, value)
+
+	configData, err := i.configFileData()
+	if err != nil {
+		return err
+	}
+
+	configData[strings.ToLower(key)] = value
+	i.data = configData
+
+	err = i.save()
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// GetString retrieves a string for a given key
-func (i *Instance) GetString(key string) string {
+func retryLock(pl *lockfile.PidLock, attempts int, interval time.Duration) error {
+	lockedErr := &lockfile.AlreadyLockedError{}
+	_, err := pl.TryLock()
+	if err != nil {
+		if !errors.As(err, &lockedErr) {
+			return errs.Wrap(err, "Unexpected error attempting to acquire lock file")
+		}
+		if attempts--; attempts > 0 {
+			time.Sleep(interval)
+			return retryLock(pl, attempts, interval)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (i *Instance) get(key string) interface{} {
 	i.rwLock.RLock()
 	defer i.rwLock.RUnlock()
```

Comment:
We can drop this now.

Reply:
I think we still need this, otherwise we could encounter a panic attempting to read from a map that is currently being written to

Reply:
How? The only way that that could happen is if a subroutine is writing to i.data, in which case it would be more appropriate to use a mutex. But I think that mutex probably makes more sense when writing than when reading.

Reply:
The scenario you describe is also the one I'm thinking of. It is unlikely to happen in the same instance of the state tool but still possible I think. The `rwLock` is a mutex which is used in both `Set` and `Get`

Reply:
A small aside: In almost all cases a RWLock costs more than the optimization it provides. In other words, it requires quite a bit of contention before it's worth using.